fdbserver/GlobalTagThrottler.actor.cpp

```diff
@@ -0,0 +1,191 @@
+/*
+ * GlobalTagThrottler.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbclient/FDBTypes.h"
+#include "fdbclient/TagThrottle.actor.h"
+#include "fdbrpc/Smoother.h"
+#include "fdbserver/TagThrottler.h"
+
+#include <limits>
+
+#include "flow/actorcompiler.h" // must be last include
+
+class GlobalTagThrottlerImpl {
+	class QuotaAndCounters {
+		Optional<ThrottleApi::TagQuotaValue> quota;
+		Smoother readCostCounter;
+		Smoother writeCostCounter;
+		Smoother transactionCounter;
+
+	public:
+		QuotaAndCounters()
+		  : readCostCounter(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_FOLDING_TIME),
+		    writeCostCounter(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_FOLDING_TIME),
+		    transactionCounter(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_FOLDING_TIME) {}
+
+		void setQuota(ThrottleApi::TagQuotaValue const& quota) { this->quota = quota; }
+
+		void addReadCost(double readCost) { readCostCounter.addDelta(readCost); }
+
+		void addWriteCost(double writeCost) { writeCostCounter.addDelta(writeCost); }
+
+		void addTransactions(int count) { transactionCounter.addDelta(count); }
+
+		Optional<ClientTagThrottleLimits> getTotalLimit() const {
+			if (!quota.present())
+				return {};
+			Optional<double> readLimit, writeLimit;
+			if (readCostCounter.smoothRate() > 0) {
+				readLimit = quota.get().totalReadQuota * transactionCounter.smoothRate() / readCostCounter.smoothRate();
```

Comment:
Nit: it took me a little effort to interpret this. We could fix that with comments, but I also think that reordering it might help here:

```(quota.get().totalReadQuota / readCostCounter.smoothRate()) * transactionCounter.smoothRate()```